# Aprendiendo y Practicanto Numpy con Ejemplo y Ejericios

A continuacion veremos algunos ejempos de funciones las cuales nos podrian llegar a ser muy utiles al momento de trabajar con Numpy esto junto a una parte de elaboracion de ejercicios los cuales nos ayudaran a entender y visualizar el funcionamiento de esas herramientas en problemas reales.

Lo primero que debemos hacer es importar las librerias que vamos a utilizar, las cuales serian Numpy y Matplotlib, esta ultima nos ayudara con los graficos de los datos.

In [49]:
import numpy as np # ponemos np como alias de numpy y poderlo llamar mas facil   
import matplotlib.pyplot as plt

## Lo primero que veremos sera el **Basic Indexer**
el cual nos permite visualizar en consola un elemento en especifico en un arreglo.

In [50]:
arr = np.array([10, 20, 30, 40, 50])
print(arr[2]) 

30


Como logramos observar el resultafo fue 30, ya que el indice 2 corresponde al tercer elemento de la lista, teniendo en cuenta de que las posiciones comienzan desdel el 0.

## Lo siguente que veremos sera el **Slicing**
el cual nos ayudara a visualizar solamente un intervalo de elementos de un arreglo

In [51]:
arr = np.array([10, 20, 30, 40, 50])
print(arr[1:4]) 

[20 30 40]


Como pudimos ver, pudimos ver como acceder a un conjunto de elementos o a un rango de elementos de un array.

## Ahora veremos los **Indexings en arreglos en 2D**.

In [52]:
arr = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(arr[1, 2]) 

6


Lo que nos quiere decir el 2D array es que en una matriz nxm podemos dar una instruccion de impresion en consola del elemento que se encuentra en la posicion 1 (filas), 2 (columnas).

## Lo siguiente que veremos sera el **Slicing en un 2D array**.

In [53]:
arr = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(arr[:2, 1:]) # Output: [[2 3] [5 6]]

[[2 3]
 [5 6]]


Aqui los que nos muestra es que nos mostrará todas las filas desde el inicio hasta la fila en posicion 2 y todas las columnas desde la columna 1 hasta el final.


## Ahora lo veremos los **Indexing pero en Tensores 3D**

In [54]:
tensor = np.array([[[1, 2, 3], [4, 5, 6]],
[[7, 8, 9], [10, 11, 12]],
[[13, 14, 15], [16, 17, 18]]])

# Access specific elements
print(tensor[1, 0, 2])

# Access an entire "slice" (2D submatrix)
print(tensor[1]) 

9
[[ 7  8  9]
 [10 11 12]]


Un tensor 3D en NumPy es básicamente un arreglo con 3 dimensiones. Puedes pensarlo como una lista de matrices (o una "caja" que contiene varias matrices 2D apiladas una encima de la otra).

Por ejemplo, el tensor 3D que utilizamos en el codigo de arriba, se refiere a una especie de una matriz con 3 sub matrices de 2x3 cada una.

La sintaxis para acceder al tensor es la siguiente:
tensor[depth, row, col]

Asi que segun lo atenrior el output que nos solto la consola anteriormente se refiere a:
tensor[1, 0, 2] → segunda matriz, primera fila, tercera columna (uotput: 9)

Como tambien podemos hacer que nos imprima todo el tensor que se encuentra en una posicion especifica.
Tensor[1] → toda la matriz 2x3 ubicada en la posicion 1 de todo el arreglo (output: [[ 7  8  9]
 [10 11 12]])

## Ahora veremos un **Slicing en Tensor 3D** (corte por secciones)

In [55]:
tensor = np.arange(27).reshape(3, 3, 3)
# Slice along the first dimension
print(tensor[1]) # Output: 3x3 submatrix
# Slice along the second dimension
print(tensor[:, 1, :]) # Output: 3x3 submatrix
# Slice along the third dimension
print(tensor[:, :, 2]) # Output: 3x3 submatrix

[[ 9 10 11]
 [12 13 14]
 [15 16 17]]
[[ 3  4  5]
 [12 13 14]
 [21 22 23]]
[[ 2  5  8]
 [11 14 17]
 [20 23 26]]


En el código anterior, se crea un tensor con np.arange(27).reshape(3, 3, 3), es decir, 3 matrices de 3x3 con valores del 0 al 26.

Cuando usamos tensor[1], accedemos a la segunda matriz completa del tensor (una submatriz 3x3). En cambio, tensor[:, 1, :] toma la segunda fila (índice 1) de todas las matrices, y las organiza en una nueva matriz 3x3. Por último, tensor[:, :, 2] selecciona la tercera columna de todas las matrices, también devolviendo una matriz 3x3.

Este tipo de slicing nos permite extraer secciones específicas del tensor de forma rápida y eficiente.



## 4D Tensor Indexing:

In [56]:
tensor = np.arange(48).reshape(2, 3, 2, 4)
# Access specific elements
print(tensor[1, 2, 1, 3]) # Output: 47
# Access a whole "block" (3D sub-tensor)
print(tensor[1, 2]) # Output: 2x4x3 sub-tensor

47
[[40 41 42 43]
 [44 45 46 47]]


Un tensor 4D en NumPy es un arreglo con cuatro dimensiones, que puedes imaginar como una colección de bloques 3D organizados en un eje extra. En el código anterior, se crea un tensor con np.arange(48).reshape(2, 3, 2, 4), lo que genera 2 bloques, cada uno con 3 matrices de 2x4, usando los números del 0 al 47.

Cuando usamos tensor[1, 2, 1, 3], accedemos a un único elemento del tensor: el que está en el segundo bloque, tercera matriz, primera fila y cuarta columna. El valor obtenido en ese punto es 47.

En cambio, con tensor[1, 2] accedemos a una subestructura completa dentro del tensor: una matriz de tamaño 2x4 que representa todas las filas y columnas de la tercera matriz del segundo bloque.

Este tipo de indexado permite trabajar con estructuras complejas como secuencias de imágenes, datos volumétricos o entradas para redes neuronales, de forma clara y ordenada.

## Slicing a 4D Tensor:

In [57]:
tensor = np.arange(48).reshape(2, 3, 2, 4)
# Slice along the first dimension

print(tensor[1]) # Output: 3x2x4 sub-tensor
# Slice along the second dimension
print(tensor[:, 2, :, :]) # Output: 2x2x4 sub-tensor
# Slice along the third dimension
print(tensor[:, :, 0, :]) # Output: 2x3x4 sub-tensor

[[[24 25 26 27]
  [28 29 30 31]]

 [[32 33 34 35]
  [36 37 38 39]]

 [[40 41 42 43]
  [44 45 46 47]]]
[[[16 17 18 19]
  [20 21 22 23]]

 [[40 41 42 43]
  [44 45 46 47]]]
[[[ 0  1  2  3]
  [ 8  9 10 11]
  [16 17 18 19]]

 [[24 25 26 27]
  [32 33 34 35]
  [40 41 42 43]]]


En el ejemplo, np.arange(48).reshape(2, 3, 2, 4) crea un tensor con 2 bloques, cada uno con 3 matrices de 2x4.

tensor[1] accede al segundo bloque completo, devolviendo un sub-tensor de forma 3x2x4.

tensor[:, 2, :, :] extrae la tercera matriz (índice 2) de cada bloque, resultando en un sub-tensor 2x2x4.

tensor[:, :, 0, :] toma la primera fila de todas las matrices en todos los bloques, formando un sub-tensor 2x3x4.

Este slicing permite acceder a partes específicas del tensor de forma rápida y estructurada.



## Matrix Addition:

In [58]:
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])
result = A + B
print(result)

[[ 6  8]
 [10 12]]


La suma de matrices en NumPy se realiza elemento a elemento, siempre que ambas matrices tengan las mismas dimensiones. En el ejemplo, A y B son matrices de 2x2, y al sumarlas (A + B), cada elemento en la misma posición se suma entre sí, dando como resultado una nueva matriz: [[6, 8], [10, 12]].

## Matrix Multiplication:

In [59]:
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])
result = np.dot(A, B)
print(result)

[[19 22]
 [43 50]]


La multiplicación de matrices en NumPy se puede hacer con np.dot() o el operador @, y sigue las reglas del álgebra lineal. En este caso, A y B son matrices de 2x2, y su multiplicación da como resultado otra matriz 2x2: [[19, 22], [43, 50]]. Cada elemento se calcula multiplicando filas de A por columnas de B y sumando los productos.

## Transpose of a Matrix:

In [60]:
A = np.array([[1, 2], [3, 4]])
result = np.transpose(A)
print(result)

[[1 3]
 [2 4]]


La transposición de una matriz en NumPy se hace con np.transpose(), lo cual invierte sus filas y columnas. En el ejemplo, la matriz A = [[1, 2], [3, 4]] se convierte en [[1, 3], [2, 4]], intercambiando filas por columnas. Es útil para operaciones como multiplicación de matrices o para alinear datos.

## Matrix Determinant and Inverse:

In [61]:
A = np.array([[1, 2], [3, 4]])
det = np.linalg.det(A)
inverse = np.linalg.inv(A)
print("Determinant:", det)
print("Inverse:\n", inverse)

Determinant: -2.0000000000000004
Inverse:
 [[-2.   1. ]
 [ 1.5 -0.5]]


En NumPy, el determinante de una matriz se calcula con np.linalg.det(), y su inversa con np.linalg.inv(). En el ejemplo, para la matriz A = [[1, 2], [3, 4]], el determinante es -2.0, y la matriz inversa es [[ -2. , 1. ], [ 1.5, -0.5 ]]. Ademas tambien hay que tener en cuenta que la matriz inversa solo existe si el determinante es distinto de cero.

## Eigenvalues y Eigenvectors:

In [62]:
A = np.array([[1, -1], [1, 3]])
eigenvalues, eigenvectors = np.linalg.eig(A)
print("Eigenvalues:", eigenvalues)
print("Eigenvectors:\n", eigenvectors)

Eigenvalues: [2.00000002 1.99999998]
Eigenvectors:
 [[-0.70710677 -0.70710679]
 [ 0.70710679  0.70710677]]


En NumPy, los valores propios (eigenvalues) y los vectores propios (eigenvectors) de una matriz se obtienen con np.linalg.eig(). En este caso, para la matriz A = [[1, -1], [1, 3]], se calculan dos valores propios y sus vectores asociados. Los eigenvalues indican cuánto se estiran o encogen los vectores, y los eigenvectors son las direcciones que no cambian bajo la transformación de la matriz. 

## Solving Linear Equations:

In [63]:
A = np.array([[2, 1], [1, 1]])
b = np.array([3, 2])
solution = np.linalg.solve(A, b)
print("Solution:", solution)

Solution: [1. 1.]


En NumPy, para resolver un sistema de ecuaciones lineales de la forma Ax = b, se usa np.linalg.solve(). En este ejemplo, A es la matriz de coeficientes y b es el vector de resultados. La función devuelve el vector x que satisface la ecuación. Para A = [[2, 1], [1, 1]] y b = [3, 2], la solución es [1.0, 1.0], lo que significa que x = 1 y y = 1 resuelven el sistema.

## Gradient descent optimization:

Dado el conjunto de datos $X$ e $y$:

$$
X = \begin{bmatrix} x(1) \\ x(2) \\ \vdots \\ x(m) \end{bmatrix}, \quad y = \begin{bmatrix} y(1) \\ y(2) \\ \vdots \\ y(m) \end{bmatrix}
$$

Donde $m$ es el número de muestras.

El modelo de regresión lineal se define como:

$$
h_{\theta}(x) = \theta_0 + \theta_1 x
$$

Añadimos un término de sesgo a las características, creando $X_b$:

$$
X_b = \begin{bmatrix} 1 & x(1) \\ 1 & x(2) \\ \vdots & \vdots \\ 1 & x(m) \end{bmatrix}
$$

El objetivo es minimizar el error cuadrático medio:

$$
J(\theta) = \frac{1}{2m} \sum_{i=1}^{m} \left( h_{\theta}(x(i)) - y(i) \right)^2
$$

El descenso de gradiente actualiza $\theta$ usando los gradientes:

$$
\theta := \theta - \alpha \nabla J(\theta)
$$

Donde $\alpha$ es la tasa de aprendizaje.

El gradiente $\nabla J(\theta)$ con respecto a $\theta$ se calcula como:

$$
\nabla J(\theta) = \frac{1}{m} X_b^T (X_b \theta - y)
$$

Iteramos este proceso durante un número específico de iteraciones.

Los parámetros optimizados finales son:

$$
\theta = \begin{bmatrix} \theta_0 \\ \theta_1 \end{bmatrix}
$$

El proceso de optimización por descenso de gradiente puede ser codificado en Python como sigue:

In [64]:
# Generate some sample data
np.random.seed(0)
X = 2 * np.random.rand(100, 1)
y = 4 + 3 * X + np.random.randn(100, 1)
# Add bias term to X
X_b = np.c_[np.ones((100, 1)), X]
# Hyperparameters
learning_rate = 0.1
n_iterations = 1000
# Initialize the parameters
theta = np.random.randn(2, 1)
# Gradient Descent
for iteration in range(n_iterations):
    gradients = 2 / 100 * X_b.T.dot(X_b.dot(theta) - y)
    theta = theta - learning_rate * gradients
print("Final theta:", theta)

Final theta: [[4.22215108]
 [2.96846751]]


Este código se implementa el descenso de gradiente para una regresión lineal simple. Primero, genera datos sintéticos: X con 100 valores aleatorios entre 0 y 2, y $y$ siguiendo la relación $y = 4 + 3X + \text{ruido}$. Luego, agrega un término de sesgo a X (creando X_b) para incluir $\theta_0$. Define hiperparámetros: una tasa de aprendizaje de 0.1 y 1000 iteraciones. Inicializa $\theta$ (los parámetros $\theta_0$ y $\theta_1$) aleatoriamente y aplica el descenso de gradiente iterativamente, calculando los gradientes como $\frac{2}{m} X_b^T (X_b \theta - y)$ y actualizando $\theta$ con la regla $\theta = \theta - \alpha \cdot \text{gradientes}$. Finalmente, imprime los valores optimizados de $\theta$.

# Ejmplos: Python funciones y clases


## Simple Function:

In [48]:
def greet(name):
    return "Hello, " + name + "!"
print(greet("Ana")) # Output: Hello, Ana!

Hello, Ana!


Este código en Python define una función simple llamada greet que toma un parámetro name. La función concatena el texto "Hello, " con el valor de name y un signo de exclamación "!", retornando el saludo completo. Luego, se llama a la función con el argumento "Ana", y al ejecutar print(greet("Ana")), se imprime el resultado: "Hello, Ana!". Es un ejemplo básico de cómo definir y usar una función en Python para generar un mensaje personalizado.

## Funcion con parametros y retornos:

In [65]:
def add_numbers(a, b):
    sum = a + b
    return sum

result = add_numbers(5, 7)
print(result) # Output: 12

12


Este código en Python define una función add_numbers que toma dos parámetros, a y b. Dentro de la función, calcula la suma de a y b, la almacena en la variable sum, y retorna ese valor. Luego, se llama a la función con los argumentos 5 y 7, y el resultado (12) se guarda en la variable result. Finalmente, print(result) imprime el valor 12. Es un ejemplo sencillo de cómo crear y usar una función para realizar una operación aritmética básica.

## Funcion con parametros por defecto:

In [67]:
def power(base, exponent=2):
    return base ** exponent
print(power(3)) # Output: 9
print(power(2, 3)) # Output: 8

9
8


Este código en Python define una función power que calcula la potencia de un número. Tiene dos parámetros: base y exponent, donde exponent tiene un valor por defecto de 2. La función retorna base elevado a exponent usando el operador **. Al llamar power(3), usa el valor por defecto de exponent, resultando en $3^2 = 9$. Al llamar power(2, 3), especifica exponent como 3, dando $2^3 = 8$. Es un ejemplo de cómo usar parámetros por defecto en funciones para hacerlas más flexibles.

## Functiones con multiplicadores y retornos:

In [68]:
def min_max(numbers):
    return min(numbers), max(numbers)
nums = [4, 9, 2, 7, 5]
minimum, maximum = min_max(nums)
print(minimum, maximum) # Output: 2 9

2 9


Este código define una función min_max que toma una lista numbers y retorna dos valores: el mínimo y el máximo de la lista, usando las funciones integradas min() y max(). La lista nums contiene [4, 9, 2, 7, 5]. Al llamar min_max(nums), la función retorna 2 (mínimo) y 9 (máximo), que se asignan a las variables minimum y maximum mediante desempaquetado. Luego, print(minimum, maximum) imprime "2 9". Es un ejemplo de cómo una función puede retornar múltiples valores y cómo desempaquetarlos en variables.

## Functiones con Docstring:

In [71]:
def area_of_rectangle(length, width):
    """
    Calculates the area of a rectangle.
    Args:
        length (float): The length of the rectangle.
        width (float): The width of the rectangle.
    Returns:
        float: The area of the rectangle.
    """
    return length * width

Este codigo define una función area_of_rectangle que calcula el área de un rectángulo. Toma dos parámetros, length (longitud) y width (ancho), ambos de tipo float según la documentación. La función multiplica length por width y retorna el resultado, que representa el área. La docstring explica el propósito de la función, los tipos de los argumentos y el valor retornado (un float). Es un ejemplo de cómo crear una función simple con documentación clara para calcular el área de un rectángulo.

## Creando una Clase Simple:

In [73]:
class Dog:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    def bark(self):
        print(f"{self.name} is barking!")
# Create an instance of the Dog class
dog1 = Dog("Buddy", 3)
dog1.bark() # Output: Buddy is barking!

Buddy is barking!


Este código en Python define una clase Dog para modelar un perro. El método __init__ es el constructor, que inicializa dos atributos: name (nombre) y age (edad), usando los parámetros proporcionados. El método bark imprime un mensaje que indica que el perro (usando self.name) está ladrando. Se crea una instancia dog1 de la clase Dog con nombre "Buddy" y edad 3. Al llamar dog1.bark(), se ejecuta el método y se imprime "Buddy is barking!". Es un ejemplo básico de programación orientada a objetos, mostrando cómo definir una clase, crear un objeto y usar sus métodos.

## Adicionando Metodos y Atributos

In [78]:
class Student:
    def __init__(self, name, age):
        self.name = name
        self.age = age
        self.grades = []
    def add_grade(self, grade):
        self.grades.append(grade)
    def get_average_grade(self):
        return sum(self.grades) / len(self.grades)
# Create a Student instance and add grades
student1 = Student("Alice", 18)
student1.add_grade(85)
student1.add_grade(92)
average_grade = student1.get_average_grade()
print(f"{student1.name}’s average grade: {average_grade}")

Alice’s average grade: 88.5


Este código en Python define una clase Student para representar un estudiante. El constructor __init__ inicializa los atributos name (nombre), age (edad) y una lista vacía grades (notas). El método add_grade agrega una nota a la lista grades, y get_average_grade calcula el promedio de las notas dividiendo la suma de las notas entre su cantidad. Se crea una instancia student1 para "Alice", de 18 años, se añaden las notas 85 y 92, y se calcula su promedio (88.5). Al ejecutar print, se muestra "Alice’s average grade: 88.5". Es un ejemplo de cómo usar clases para gestionar datos y calcular estadísticas.

## Inheritance

In [80]:
class Animal:
    def __init__(self, name):
        self.name = name
    def speak(self):
        pass # Placeholder for subclasses to override
class Cat(Animal):
    def speak(self):
        print(f"{self.name} says Meow!")
class Dog(Animal):
    def speak(self):
        print(f"{self.name} says Woof!")
# Create instances of Cat and Dog
cat = Cat("Whiskers")
dog = Dog("Buddy")
cat.speak() # Output: Whiskers says Meow!
dog.speak() # Output: Buddy says Woof!

Whiskers says Meow!
Buddy says Woof!


Este código en Python demuestra la herencia en programación orientada a objetos. La clase base Animal tiene un constructor que inicializa el atributo name y un método speak vacío como marcador. Las clases Cat y Dog heredan de Animal y sobrescriben el método speak para que cada una imprima un sonido específico: "Meow" para Cat y "Woof" para Dog. Se crean instancias cat ("Whiskers") y dog ("Buddy"), y al llamar speak() en cada una, se imprime "Whiskers says Meow!" y "Buddy says Woof!". Es un ejemplo de cómo la herencia permite reutilizar código y especializar comportamientos en subclases.

# EJERCICIOS

## 3.1 Indexing Arrays in NumPy
1. Dado el arreglo de NumPy `arr = np.array([3, 7, 1, 9, 5])`, realiza las siguientes operaciones:

   - Recupera el tercer elemento del arreglo (índice 2 a índice 3). 
   - Selecciona los elementos desde el índice 1 hasta el índice 3 (sin incluir el índice 3). 
   - Cambia el valor del cuarto elemento a 12.  

In [96]:
arr = np.array([3, 7, 1, 9, 5])
print(arr[2]) 
print(arr[1:4])
arr[3] = 12
print(arr[3])

1
[7 1 9]
12


2. Crea una matriz 2D en NumPy con los siguientes valores:

$$
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{bmatrix}
$$

   a) Recupera el elemento en la segunda fila y tercera columna.  
   b) Corta la matriz para obtener la submatriz que consiste en las primeras dos filas y todas las columnas.

In [95]:
arr = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(arr[1, 2]) # Output: 6
print(arr[:2, 0:]) # Output: [[2 3] [5 6]]

6
[[1 2 3]
 [4 5 6]]


## 3.2 Indexing and Slicing Higher-Dimensional Tensors

Considera el siguiente tensor 3D que representa un conjunto de matrices:

$$
\text{tensor} = \begin{bmatrix}
\begin{bmatrix} 1 & 2 & 3 & 4 \\ 5 & 6 & 7 & 8 \\ 9 & 10 & 11 & 12 \end{bmatrix}, \quad
\begin{bmatrix} 13 & 14 & 15 & 16 \\ 17 & 18 & 19 & 20 \\ 21 & 22 & 23 & 24 \end{bmatrix}, \quad
\begin{bmatrix} 25 & 26 & 27 & 28 \\ 29 & 30 & 31 & 32 \\ 33 & 34 & 35 & 36 \end{bmatrix}
\end{bmatrix}
$$

1. Recupera el elemento en la posición (1, 0, 2) dentro del tensor.  
2. Corta el tensor para obtener el sub-tensor 3 x 4 que corresponde a la segunda "matriz" (slice) a lo largo de la primera dimensión.  
3. Corta el tensor para obtener el subtensor 2 x 4 que corresponde a la última fila de las dos últimas matrices.
4. Recupera el elemento en la posición (2, 1, 3) dentro del tensor.  
5. Corta el tensor para obtener el subtensor 2 x 4 que corresponde a la primera fila de las dos últimas matrices.

**Nota:** Recuerda que en la indexación de NumPy, el primer índice corresponde a la dimensión más externa, el segundo índice a la dimensión interna siguiente, y así sucesivamente.

In [112]:
tensor = np.array([[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]],
                   [[13, 14, 15, 16], [17, 18, 19, 20], [21, 22, 23, 24]],
                   [[25, 26, 27, 28], [29, 30, 31, 32], [33, 34, 35, 36]]])

print("1--> ", tensor[1, 0, 2])
print("2--> ", tensor[1])
print("3--> ", tensor[1:, 2, :])
print("4--> ", tensor[2, 1, 3])
print("5--> ", tensor[:2, 0, :])

1-->  15
2-->  [[13 14 15 16]
 [17 18 19 20]
 [21 22 23 24]]
3-->  [[21 22 23 24]
 [33 34 35 36]]
4-->  32
5-->  [[ 1  2  3  4]
 [13 14 15 16]]


## 3.3 Defining Functions in Python

1. Escribe una función Python calcular promedio que tome una lista de números como entrada y devuelva el promedio de esos números.
2. Define una función encontrar potencia que calcule el resultado de un número elevado a un exponente dado. La función debe tener dos parámetros: base (el
número base) y exponente (el valor por defecto debe ser 2).